### PR TITLE
[CARBONDATA-3244] Add benchmark for Change Data Capture scenario

### PR DESCRIPTION
**Pull request by jackylk**
CDC (change data capture) is a common scenario for analyzing slowly changed table in data warehouse.
It is good to add benchmark test comparing two update methods:
1. hive_solution, which uses INSERT OVERWRITE. This is a popular method for hive warehouse.
2. carbon_solution, which uses CarbonData's update syntax to update the history table directly.

This test simulates updates to history table using CDC table.
When running in a 8-cores laptop, the benchmark shows:
1. test one
History table 1M records, update 10K records everyday and insert 10K records everyday, simulated 3 days.
hive_solution: total process time takes 13,516 ms
carbon_solution: total process time takes 7,521 ms

2. test two
History table 10M records, update 10K records everyday and insert 10K records everyday,
simulated 3 days.
hive_solution: total process time takes 104,250 ms
carbon_solution: total process time takes 17,384 ms

 - [X] Any interfaces changed?
 No
 - [X] Any backward compatibility impacted?
 No
 - [X] Document update required?
No
 - [X] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
  Only example is added     
 - [X] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
NA

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2258/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2477/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10516/

File: examples/spark2/src/main/scala/org/apache/carbondata/benchmark/CDCBenchmark.scala [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+      if (printDetail) {
+        println(s"day$i CDC")
+        spark.sql("select * from ods_order").show(100, false)
+      }
+
+      // update dw table using CDC data
+      val start = System.nanoTime()
+      hive_solution(spark)
+      // carbon_solution(spark)
```
Comment (qiuchenjian) [inline examples/spark2/src/main/scala/org/apache/carbondata/benchmark/CDCBenchmark.scala]: Why is carbon_solution commented, if either hive_solution or carbon_solution is tested once, better to add a flag to control in "args: Array[String]"

**Comment by qiuchenjian** [comment]
i think the query performance of carbon_solution is lower than hive_solution's, because carbon_solution has more segment (insert generates a segment and update generates more segment)
Do we have some method to optimize this?

File: examples/spark2/src/main/scala/org/apache/carbondata/benchmark/CDCBenchmark.scala (hunk 1 above)
Comment (qiuchenjian) [inline examples/spark2/src/main/scala/org/apache/carbondata/benchmark/CDCBenchmark.scala]: better to add a variable to run hive_solution or carbon_solution separately， using comment is confused to others

File: examples/spark2/src/main/scala/org/apache/carbondata/benchmark/CDCBenchmark.scala [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+  }
+
+  /**
+   * Typical solution when using hive
+   * This solution uses INSERT OVERWRITE to rewrite the whole table every day
+   */
+  private def hive_solution(spark: SparkSession) = {
+    spark.sql(
+      """
+        | insert overwrite table dw_order
```
Comment (ravipesala) [inline examples/spark2/src/main/scala/org/apache/carbondata/benchmark/CDCBenchmark.scala]: Maybe it is better to create a partition table on date column and insert overwrite only on the partition would be better.

**Comment by ravipesala** [comment]
> i think the query performance of carbon_solution is lower than hive_solution's, because carbon_solution has more segment (insert generates a segment and update generates more segment)
> Do we have some method to optimize this?

Since we are updating the existing data it creates extra files like delete delta and incremental carbondata files It may degrade query performance a little but when do the compaction it will get improved.

**Comment by ravipesala** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2607/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10866/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2838/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/36/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/35/

File: examples/spark2/src/main/scala/org/apache/carbondata/benchmark/CDCBenchmark.scala (hunk 2 above)
Comment (jackylk) [reply examples/spark2/src/main/scala/org/apache/carbondata/benchmark/CDCBenchmark.scala]: I will do it in another PR.

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/780/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/786/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/788/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/781/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/787/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/789/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/798/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/803/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/806/

**Comment by kunal642** [comment]
LGTM

**Pull request closed without merging**